domain/src/main/java/org/oppia/android/domain/state/StateDeck.kt

```diff
@@ -206,15 +266,33 @@ internal class StateDeck internal constructor(
           }.build()
         }
       )
+
       pendingStateName = pendingTopState.name
-      hintIndex = revealedHintIndex
       addAllPendingUserAnswers(currentDialogInteractions)
-      this.solutionIsRevealed = this@StateDeck.solutionIsRevealed
       this.stateIndex = this@StateDeck.stateIndex
+
+      hintCount = pendingTopState.interaction.hintCount
+      indexOfLastRevealedHint = revealedHintsAndSolutionIndex
+      this.hintIndex =
+        if (isUnrevealedHintVisible) {
+          revealedHintsAndSolutionIndex + 1
+        } else {
+          revealedHintsAndSolutionIndex
+        }
+      this.isUnrevealedSolutionVisible = this@StateDeck.isUnrevealedSolutionVisible
+
```

Comment:
This method of creating checkpoints is becoming over complicated. I am thinking of having separate variables for hint and solution (2 for each)  
--  revealed hint/solution index: Int
--  unrevealed hint/solution Index: Int

WDYT?

Reply:
Could we maybe leverage HelpIndex for this purpose? It was originally meant to help simplify this mechanism in the UI layer, but it seems like it could be used here as well.

Reply:
resolving this because changed the approach to handle hints